### DISCUSSION TITLE
What are the arguments and type arguments need to be passed in verify_proof function from module groth16_example::groth16 ?

### DISCUSSION THREAD

**Discussion by mohan00710** [aptos-labs/aptos-developer-discussions#436] [Questions]
### Discord user ID

cypher0343

### Describe your question in detail.

I have generated a ZK proof using snarkjs. How can I verify the generated proof on-chain in Aptos using the verify_proof function?

https://github.com/aptos-labs/aptos-core/blob/574fa48194ca42013b4b5171ec81ebfb5ea5ac14/aptos-move/move-examples/groth16_example/sources/groth16.move#L14

for example this is the proof that i generated

```
 proof: {
  pi_a: [
    '16286470799066585029045712586541356264760704920573070700140933006710149896371',
    '833307193702532317725103948360422327062981515765809063874415038203096058937',
    '1'
  ],
  pi_b: [
    [
      '880926228059373207667992619774665046942965620149950199919426317382177917661',
      '6140866589518306847019528203961484763388155655750788241176276716882314745815'
    ],
    [
      '6409412076291721770121775687361519485536726545492613357801511798281987008969',
      '7751893097262400962491938395348268043987382887839762573492579241342775286031'
    ],
    [ '1', '0' ]
  ],
  pi_c: [
    '16110752813674205563490857085332825743371665235634081447638995489622937222272',
    '18997759920270420094929115777609656804698138158035252675022978640238051592809',
    '1'
  ],
  protocol: 'groth16',
  curve: 'bn128'
}
```

### Which operating system are you using?

macOS

### Which SDK or tool are you using? (if any)

TypeScript SDK

**Comment by zjma** [comment]
Take a look at [this unit test](https://github.com/aptos-labs/aptos-core/blob/574fa48194ca42013b4b5171ec81ebfb5ea5ac14/aptos-move/move-examples/groth16_example/sources/groth16.move#L87) which demonstrates the usage. It uses bls12381 types (defined [here](https://github.com/aptos-labs/aptos-core/blob/574fa48194ca42013b4b5171ec81ebfb5ea5ac14/aptos-move/framework/aptos-stdlib/sources/cryptography/bls12381.move)) but you can easily modify it to bn254 types (defined [here](https://github.com/aptos-labs/aptos-core/blob/574fa48194ca42013b4b5171ec81ebfb5ea5ac14/aptos-move/framework/aptos-stdlib/sources/cryptography/bn254_algebra.move#L164)), which is your case?

Note that G1/G2 elements that snarkjs gives you in proof/verification keys can be in projective coordinates `(decimal_x, decimal_y, decimal_z)` and you need to do the following.
1. convert them to affine coordinates.
   - Typically the bn254 library you use should have supported it.
1. serialize them into bytes using format equivalent to what aptos API supports ([FormatG1Uncompr](https://github.com/aptos-labs/aptos-core/blob/574fa48194ca42013b4b5171ec81ebfb5ea5ac14/aptos-move/framework/aptos-stdlib/sources/cryptography/bn254_algebra.move#L141)/[FormatG1Compr](https://github.com/aptos-labs/aptos-core/blob/574fa48194ca42013b4b5171ec81ebfb5ea5ac14/aptos-move/framework/aptos-stdlib/sources/cryptography/bn254_algebra.move#L164)/[FormatG2Uncompr](https://github.com/aptos-labs/aptos-core/blob/574fa48194ca42013b4b5171ec81ebfb5ea5ac14/aptos-move/framework/aptos-stdlib/sources/cryptography/bn254_algebra.move#L193)/[FormatG2Compr](https://github.com/aptos-labs/aptos-core/blob/574fa48194ca42013b4b5171ec81ebfb5ea5ac14/aptos-move/framework/aptos-stdlib/sources/cryptography/bn254_algebra.move#L216))
   - Tyipcally the bn254 library you use serializes/compresses G1 points using the format equivalent to `FormatG1Compr` (and `FormatG2Compr` for G2 points). But better read their documentation/write a test to confirm, there seems no universally adopted format.
1. call the deserialize function with the actual format you use in the previous step to get type-safe elliptic curve point objects in Move.
1. call the verify function with these point objects.

Let me know if you need more details.
